// @iota/wallet/out/lib/bindings.js
"use strict";
// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.destroy = exports.clearListeners = exports.listen = exports.messageHandlerNew = exports.sendMessageAsync = exports.internalInitLogger = void 0;
// @ts-ignore: path is set to match runtime transpiled js path
const addon = require("../../build/Release/index.node");
const { initLogger, sendMessage, messageHandlerNew, listen, clearListeners, destroy, } = addon;
exports.internalInitLogger = initLogger;
exports.messageHandlerNew = messageHandlerNew;
exports.listen = listen;
exports.clearListeners = clearListeners;
exports.destroy = destroy;
const sendMessageAsync = (message, handler) => new Promise((resolve, reject) => {
    sendMessage(message, handler, (error, result) => {
        if (error) {
            reject(error);
        }
        else {
            resolve(result);
        }
    });
});
exports.sendMessageAsync = sendMessageAsync;
